Implement the federation API server initialization code in Go in order to make it more testable.

madhusudancs opened the pull request:
This initialization code is expected to be run inside an init container in the federation API server pod. Only a very basic test exists now, just to prove the point and as a sanity check. I am working on additional tests.

There is one more component missing. I need to generate a secret specific to admin client's kubeconfig. That logic is in works.

Also, I think the code is self-explanatory. Please let me know if it isn't.

``` release-note
Federation API server initialization code is now implemented in Go and is expected to be run inside an init container in the federation API server pod.
```

cc @kubernetes/sig-cluster-federation @colhom

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30615)

<!-- Reviewable:end -->

madhusudancs committed:
Initial implementation of federation API server initialization code.

madhusudancs committed:
No need to base64 encode PEM bytes for secrets, secrets automatically does that.

madhusudancs committed:
Use the CA certificates and private key to sign the server and client certificates.

madhusudancs committed:
Refactor to pass IP addresses with a stronger type.

madhusudancs committed:
Describe the overall logic of this initialization program in the package-level doc.

madhusudancs committed:
Generate controller manager kubeconfig.

madhusudancs committed:
Populate federation API server's advertise-address in the main credentials secret.

madhusudancs committed:
Refactor certs function to re-use the function that provides canonical address/names.

madhusudancs committed:
Read the federation controller manager kubeconfig from a filesystem path.

This decoupling from the Kubernetes API allows admins to run federation
control plane components wherever they like, even outside Kubernetes.

madhusudancs committed:
Generate a token and a password for an admin account.

This is purely for convenience. Tokens could be used in command line
clients such as `curl` with the HTTP header `Authorization: Bearer
<TOKEN>`. Password helps browsers to present a HTTP basic auth challenge
and filling in username and password helps the users to access the API
server from the browser.

madhusudancs committed:
Refactor the code a bit to remove unnecessary buffer usage.

madhusudancs committed:
Implement the initial set of TLS certificates test.

nikhiljindal commented on federation/deploy/init/dep.yaml:
```diff
@@ -0,0 +1,18 @@
+kind: Deployment
+metadata:
+  labels:
+    run: fi
```
Did you mean federation-init by fi? We encourage clearer labels rather than shorter ones.

nikhiljindal commented on federation/deploy/init/dep.yaml:
```diff
@@ -0,0 +1,18 @@
+metadata:
+  labels:
+    run: fi
+  name: fi
```
Put it in federation namespace? (Same as other federation pods)

nikhiljindal commented on federation/deploy/init/init.go:
```diff
@@ -0,0 +1,573 @@
+// create these secrets might mix up values from different initializers
+// leading to invalid credentials. We could elect a leader to decide which of
+// these initializers are allowed to create those credentials, but that isn't
+// sufficient since we also need guards against pod restarts. So instead, we
```
Sorry did not understand this. We can elect a new leader in case of pod restart?

nikhiljindal commented on federation/deploy/init/init.go:
```diff
@@ -0,0 +1,573 @@
+// After creating all the credentials in memory, it attempts to write them to
+// the secret indicated by the flag. If that fails it successfully exits.
+// Otherwise it continues to write all the other secrets. Effectively, the
+// first secret, i.e. the secret indicated by the `--secret` flag acts as
```
What if the container dies after creating the first secret? Rest of the secrets will never be created?

nikhiljindal commented on federation/deploy/init/init.go:
```diff
@@ -0,0 +1,573 @@
+	namespace    = flag.String("namespace", "federation", "namespace of the federation control plane components")
+	secretName   = flag.String("secret", "federation-apiserver-credentials", "name of the secret where the federation API server user credentials are stored")
+	cmSecretName = flag.String("controllermanager-kubeconfig-secret", "federation-controller-manager-kubeconfig", "name of the secret where the federation controller manager kubeconfig is stored")
+	svcName      = flag.String("service", "federation-apiserver", "namespace of the federation control plane components")
```
copy paste error in svcName description

nikhiljindal commented on federation/deploy/init/init.go:
```diff
@@ -0,0 +1,573 @@
+		return nil, fmt.Errorf("failed to generate a server certificate: %v", err)
+	}
+
+	// TODO(madhusudancs): Ideally these clients (Controller Manager and Admin
```
Sorry what do you meany by "Admin kubconfig"?

nikhiljindal commented on federation/deploy/init/Makefile:
```diff
@@ -0,0 +1,33 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+IMAGE_NAME?=gcr.io/madhusudancs-containers/federation/init
```
Change to gcr.io/google_containers?

nikhiljindal commented on federation/deploy/init/init.go:
```diff
@@ -0,0 +1,573 @@
+*/
+
+// This program implements the initialization code required to start a
+// federation API server. This code is intended to be run inside an init
```
Will we have such init containers for other components, like one for federation-controller-manager as well.
Should we name this one federation-apiserver-init.go in that case?

nikhiljindal commented on federation/deploy/init/init.go:
```diff
@@ -0,0 +1,573 @@
+
+// This program implements the initialization code required to start a
+// federation API server. This code is intended to be run inside an init
+// container in the federation API server pod.
```
If there is some documentation of init container somewhere, will be great if you can link to it here.

madhusudancs committed:
For backwards compatibility, directly try to build the rest client config from the secret if --kubeconfig flag is empty.

quinton-hoole commented:
@madhusudancs @nikhiljindal Is this likely to merge today/tomorrow?

madhusudancs commented:
@quinton-hoole this is P1.5 for v1.4. Also, there isn't anything that is using this code right now. It is nice to get this in, but if it doesn't it is likely going to become P0 for v1.5. So I am fine dropping this for v1.4. I will prioritize it for v1.5.

quinton-hoole commented:
Thanks @madhusudancs

madhusudancs committed:
Address review comments.

madhusudancs commented:
@nikhiljindal Addressed your review comments. PTAL.

---

Review status: 0 of 6 files reviewed at latest revision, 9 unresolved discussions.

---

*[federation/deploy/init/init.go, line 18 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkZ-r1-18:-KQNQjY4G_mOBOMk-ILP:b-35dkx) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/init.go#L18)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Will we have such init containers for other components, like one for federation-controller-manager as well.
> 
> Should we name this one federation-apiserver-init.go in that case?
</details>
This is an init for federation as a whole. Future inits can qualify their names with prefixes and/or suffixes.

---

*[federation/deploy/init/init.go, line 19 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkZ-r1-19:-KQNRUy4UhlnyQg8N9P8:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/init.go#L19)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> If there is some documentation of init container somewhere, will be great if you can link to it here.
</details>
Done.

---

*[federation/deploy/init/init.go, line 30 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkZ-r1-30:-KQNRXgoQhOk4YonxuoA:b-3yw4j2) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/init.go#L30)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Sorry did not understand this. We can elect a new leader in case of pod restart?
</details>
We shouldn't. If a pod restarts the init container should not be run if all the credentials were already generated. Electing a leader and running the init container regenerates all the credentials. And that is bad.

---

*[federation/deploy/init/init.go, line 40 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkZ-r1-40:-KQNRwEqAdR6a4yiy5fy:b-7k62s2) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/init.go#L40)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> What if the container dies after creating the first secret? Rest of the secrets will never be created?
</details>
Yeah, I am aware of this problem. We need something akin to two-phase commit. I will send that in a follow up PR (not this one). Thanks for pointing it out though.

---

*[federation/deploy/init/init.go, line 94 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkZ-r1-94:-KQNSQBC20SuL0v3-mFF:btmlx8w) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/init.go#L94)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> copy paste error in svcName description
</details>
Thanks for catching that! Fixed.

---

*[federation/deploy/init/init.go, line 333 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkZ-r1-333:-KQNSTp-TWCqxJ40rpn7:b-6scctw) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/init.go#L333)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Sorry what do you meany by "Admin kubconfig"?
</details>
All the kubeconfigs that we generate today are for admin user. Each user will have their own kubeconfig, this is for the federation admin.

---

*[federation/deploy/init/Makefile, line 15 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkX-r1-15:-KQNTcw2ep4QmgL11Mjt:bb1rn1d) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/Makefile#L15)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Change to gcr.io/google_containers?
</details>
Changed this to `gcr.kubernetes.io` (Our new name going forward).

---

*[federation/deploy/init/dep.yaml, line 5 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkY-r1-5:-KQNPdyMz_-mw-E4gosz:bcbro82) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/dep.yaml#L5)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Did you mean federation-init by fi? We encourage clearer labels rather than shorter ones.
</details>
This file shouldn't be here. I might have accidentally added it via `git add .`. I was just using it for testing. Removing this file.

---

*[federation/deploy/init/dep.yaml, line 6 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-KPCM2M-jFZVooZMXqkY-r1-6:-KQNPeOr98tWzItExVc2:b2nreh6) ([raw file](https://github.com/kubernetes/kubernetes/blob/e13f0ba9ee9eefed396eee38760d0206f2350c23/federation/deploy/init/dep.yaml#L6)):*
<details><summary><i>Previously, nikhiljindal (Nikhil Jindal) wrote…</i></summary>
> Put it in federation namespace? (Same as other federation pods)
</details>
Same as above.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/30615#-:-KQNTj13-b6pqtJ-V1RZ:b-ttbc61)*
<!-- Sent from Reviewable.io -->

nikhiljindal commented:
Jenkins verification failed:
```
Some packages passed golint but are not listed in hack/.linted_packages.
Please add them in alphabetical order:

  echo federation/deploy/init >> hack/.linted_packages
  LANG=C sort -o hack/.linted_packages hack/.linted_packages

You can test via this script and commit the result.

FAILED   hack/make-rules/../../hack/verify-golint.sh
```

quinton-hoole commented:
Assigning to @madhusudancs to resolve verification issues.  Feel free to assign back for final review once that's fixed.

madhusudancs commented:
These tests and the feature it is testing are now implemented in kubefed. Closing this PR. I will reopen it if necessary.
